### PR TITLE
Feature/frontend validations

### LINKED ISSUE #71: Add validations to the frontend

### PR DESCRIPTION
**Pull request by perkynades**
Closes #71

### CODE REVIEW

File: apps/api/src/services/UserService.ts [hunk 1]
```diff
@@ -78,12 +80,28 @@ export class UserService extends BaseService<User> {
     }
 
     async create(dto: User): Promise<User | undefined> {
+        const alreadyExists = await this._userRepository.findOne({
+            where: { email: dto.email }
+        });
```
Comment (amundlrohne) [inline apps/api/src/services/UserService.ts]: Can this not be done with setting the email field to unique? 

https://typeorm.io/#entities/column-options

File: apps/api/src/services/UserService.ts [hunk 2]
```diff
@@ -78,12 +80,28 @@ export class UserService extends BaseService<User> {
...
+
         const user = await this._userRepository.save(dto);
 
         return user;
     }
 
     async update(id: number, dto: User): Promise<User | undefined> {
+        const alreadyExists = await this._userRepository.findOne({
+            where: { email: dto.email }
+        });
```
Comment (amundlrohne) [inline apps/api/src/services/UserService.ts]: Unique column?

File: apps/frontend/src/app/components/misc/validation/ErrorField.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import React, { ReactElement } from 'react';
+import style from './ErrorField.module.scss';
+
+export default function ErrorField({errorMsg}: {errorMsg: string}): ReactElement {
```
Comment (amundlrohne) [inline apps/frontend/src/app/components/misc/validation/ErrorField.tsx]: Do we need two components for this functionality? Couldn't the list just render one item if that is what is provided?

File: apps/frontend/src/app/components/project/ProjectEdit.tsx [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
-
+import { Button, Label, Textarea, TextField } from '@navikt/ds-react';
+import { Datepicker } from '@navikt/ds-datepicker';
+import '@navikt/ds-datepicker/lib/index.css';
 import style from './Project.module.scss';
 import ProjectTeam from '../projectTeam';
+import { format, toDate } from 'date-fns';
+import { DatepickerValue } from '@navikt/ds-datepicker/lib/Datepicker';
+import { IInsightPojectErrors } from '../../../validation/insightPoject/IInsightProjectErrors';
+import { validateInsightProject } from '../../../validation/insightPoject/validateInsightProject';
```
Comment (amundlrohne) [inline apps/frontend/src/app/components/project/ProjectEdit.tsx]: Should probably get some treatment in the form of a barrel file :)

File: apps/frontend/src/app/components/project/ProjectEdit.tsx [hunk 5]
```diff
@@ -18,13 +25,36 @@ export const ProjectEdit = ({
...
     loading,
 }: IProps): ReactElement => {
+    const [errorMessages, setErrorMessages] = useState<IInsightPojectErrors>(
+        {
+            nameErrorMsg: '',
+            descriptionErrorMsg: '',
+            datesErrorMsg: [],
+            projectTeamErrorMsg: ''
+        }
+    );
```
Comment (amundlrohne) [inline apps/frontend/src/app/components/project/ProjectEdit.tsx]: A general comment: there will quickly become an abundance of snippets like this throughout the code, with small differences. I think validation might be a prime candidate for the Context API.

File: apps/frontend/src/app/containers/createInsight/CreateInsight.tsx [hunk 6]
```diff
@@ -40,43 +43,67 @@ export const CreateInsight = (): ReactElement => {
...
         const payload = {
             ...insight,
             project: { ...insight.project, id: parseInt(id) }
         };
 
-        const insightMutation = await createInsight(payload);
-
-        if (!insightMutation.response || insightMutation.isError) throw new Error('Failed to post insight.');
+        // I know this "any" is dirty, but i could not find insightMutation's type
+        let insightMutation: any = undefined;
```
Comment (amundlrohne) [inline apps/frontend/src/app/containers/createInsight/CreateInsight.tsx]: All mutation hooks return an object with the following format: 

`{response: Type, isLoading: boolean, isError: AxiosError}`

Unfortunately this is not defined as an interface, maybe it should be. :P

File: apps/frontend/src/app/containers/project/CreateInsightProject.tsx [hunk 7]
```diff
@@ -25,6 +24,8 @@ export const CreateInsightProject = (): ReactElement => {
     const handleSubmit = async (project: IInsightProject) => {
         const { response, isLoading, isError } = await createInsightProject(project);
 
+        //isInsightProjectValid(insightProject).isValid
```
Comment (amundlrohne) [inline apps/frontend/src/app/containers/project/CreateInsightProject.tsx]: Something wrong here?

File: apps/frontend/src/app/containers/userProfile/UserEditProfile.tsx [hunk 8]
```diff
@@ -23,15 +30,24 @@ export const UserEditProfile = ({ originalUser, toggleEdit }: IProps): ReactElem
...
+            setPatching(true);
+            const { response, isError } = await updateUser({ ...user, latestUpdate: Date.now().toString() });
+            if (response) {
+                toggleEdit();
+            } else if (isError?.response?.status === 406) {
+                setErrorMessages({
+                    nameErrorMsg: errorMessages.nameErrorMsg,
+                    emailErrorMsg: 'En bruker med denne eposten finnes allerede',
+                    phoneErrorMsg: errorMessages.phoneErrorMsg
+                });
```
Comment (amundlrohne) [inline apps/frontend/src/app/containers/userProfile/UserEditProfile.tsx]: Does the "unique" error not get picked up the same way as "empty field"? Confused as to why it's handled differently :P

File: apps/frontend/src/validation/criteria/ICriteriaErrors.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export interface ICriteriaErrors {
+    nameErrorMsg: string
+}
```
Comment (amundlrohne) [inline apps/frontend/src/validation/criteria/ICriteriaErrors.ts]: As mentioned in my general comment, would be nice to make this functionality generic.

File: apps/frontend/src/validation/utils/isEmail.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+export default function isEmail(email: string) {
+    const emailRegex =
+        /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (amundlrohne) [inline apps/frontend/src/validation/utils/isEmail.ts]: Scary! :O 

Should we not use a library for this? Not entirely sure what the convention in NAV is

File: libs/api-interfaces/src/lib/api-interfaces.ts [hunk 11]
```diff
@@ -47,7 +47,7 @@ export interface IInsightProject {
     name: string;
     description: string;
     members: IUser[];
-    start: string;
+    start: string
```
Comment (amundlrohne) [inline libs/api-interfaces/src/lib/api-interfaces.ts]: ? xD

**Review by amundlrohne — CHANGES_REQUESTED** [review]
A lot of code, good work! :D 

I have left some comments on things I think should be improved (for readability), as well as some suggestions on areas that should be explored (read: context).

File: apps/frontend/src/validation/criteria/ICriteriaErrors.ts (hunk 9 above)
Comment (perkynades) [reply apps/frontend/src/validation/criteria/ICriteriaErrors.ts]: Agree, i just had to get an approval on the core functionalities before i start refactoring

File: apps/frontend/src/app/components/misc/validation/ErrorField.tsx (hunk 3 above)
Comment (perkynades) [reply apps/frontend/src/app/components/misc/validation/ErrorField.tsx]: Agree, i fix

File: apps/frontend/src/app/containers/userProfile/UserEditProfile.tsx (hunk 8 above)
Comment (perkynades) [reply apps/frontend/src/app/containers/userProfile/UserEditProfile.tsx]: The unique error stems from a user with the same email being in the database, while the empty whield stems from the user not entering an input

Comment (perkynades) [reply apps/frontend/src/app/containers/userProfile/UserEditProfile.tsx]: I figure that a backend call has to be made to check for duplicate emails, then its more clean to use the already existing logic

Comment (amundlrohne) [reply apps/frontend/src/app/containers/userProfile/UserEditProfile.tsx]: Aight 👍

File: apps/frontend/src/validation/insightPoject/validateInsightProject.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+import { validateStartEndDates } from '../validateStartEndDates';
+
+export function validateInsightProject(
+    insightProject: IInsightProject
+): { 
+    isValid: boolean, 
+    errorMesseges: IErrorMessage 
+} {
+    let isValid = true;
+    const errorMesseges: IErrorMessage = {
```
Comment (amundlrohne) [inline apps/frontend/src/validation/insightPoject/validateInsightProject.ts]: typo :P

**Review by amundlrohne — APPROVED** [review]
Looks very good! 

Found a small typo but other than that this should be good to go! :+1:

**Pull request merged**